crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,248 @@
+            );
+
+            ReadAsDataUrl {
+                inner: self.inner,
+                _closure: closure,
+                receiver: rx,
+            }
+        }
+
+        pub fn read_as_array_buffer(self, blob: &impl BlobLike) -> ReadAsArrayBuffer {
+            let (closure, rx) = self.read_as(
+                blob,
+                std::convert::Into::into,
+                web_sys::FileReader::read_as_array_buffer,
+            );
+
+            ReadAsArrayBuffer {
+                inner: self.inner,
+                _closure: closure,
+                receiver: rx,
+            }
+        }
+
+        fn read_as<T, F, G>(
+            &self,
+            blob: &impl BlobLike,
+            convert: F,
+            start_read: G,
+        ) -> (
+            wasm_bindgen::closure::Closure<FnMut() + 'static>,
+            futures::sync::oneshot::Receiver<T>,
+        )
+        where
+            T: 'static,
+            F: Fn(wasm_bindgen::JsValue) -> T + 'static,
+            G: Fn(&web_sys::FileReader, &web_sys::Blob) -> Result<(), wasm_bindgen::JsValue>,
+        {
+            let (tx, rx) = futures::sync::oneshot::channel();
+            let reader = self.inner.clone();
+            let closure = Closure::once(move || {
+                let _ = reader.result().map(|js_value| {
+                    let _ = tx.send(convert(js_value));
+                });
+            });
+            let function = closure.as_ref().dyn_ref().unwrap_throw();
+            self.inner.clone().set_onload(Some(&function));
+            start_read(&self.inner, blob.as_raw()).unwrap_throw();
+            (closure, rx)
+        }
+    }
+
+    macro_rules! future_impls {
+       ($($name:ident : $item:ty),*) => ($(
+            pub struct $name {
+                receiver: oneshot::Receiver<$item>,
+                _closure: Closure<FnMut()>,
+                inner: web_sys::FileReader,
+            }
+
+            impl Future for $name {
+                type Item = $item;
+                type Error = FileReadError;
+
+                fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+                    match self.receiver.poll() {
+                        Ok(Async::Ready(value)) => Ok(Async::Ready(value)),
+                        Ok(Async::NotReady) => Ok(Async::NotReady),
+                        Err(e) => Err(e.into()),
+                    }
+                }
+            }
+
+            impl std::ops::Drop for $name {
+                fn drop(&mut self) {
+                    if self.inner.ready_state() < 2 {
+                        self.inner.abort();
+                    }
+                }
+            }
+       )*)
+    }
+
+    future_impls! { ReadAsString : String, ReadAsDataUrl: String, ReadAsArrayBuffer: js_sys::ArrayBuffer }
+}
+
+#[derive(Debug)]
+pub enum FileReadError {
+    AbortedEarly,
+    JsError(wasm_bindgen::JsValue),
+}
+impl std::convert::Into<FileReadError> for wasm_bindgen::JsValue {
```

Comment:
This feels extremely broad. Since there are a limited number of errors which can be thrown, we shouldn't allow for converting anything into a `FileReadError`.